### ISSUE TITLE
Extend role rowheader & row

### ISSUE THREAD

**Issue by chrislachance** [w3c/aria#1840]
Running into issue where the rowheader content needs to span more than one column, in order to maintain a unique context.
Additionally, must be able to be detached from the content within the cell being entered.

Simple example: First name column and Last name column aren't necessarily unique unto themselves, but the combination of the two can provide more context to data within that row.

More complex example: First column is a row selection column, second is unique column data. Placing rowheader on column 2 isn't universally applied to the whole row in some screen readers, which kind of makes sense (NVDA assumes the header starts at column 2, JAWS does all items in row)

Proposal: aria-label and aria-labelledby on a scope/role=row would be readback also as a rowheader

Example (each TR would be equivalent):
<tr>
  <td role="rowheader">Student A</td>
  <td>70</td>
  <td>80</td>
</tr>
<tr role="row" aria-labelledby="stud1">
  <td id="stud1">Student A</td>
  <td>70</td>
  <td>80</td>
</tr>
<tr role="row" aria-label="Student A">
  <td id="stud1">Student A</td>
  <td>70</td>
  <td>80</td>
</tr>

**Comment by chlane** [comment]
Hi @chrislachance, for clarification, are you asking [implicit roles](**[url](https://www.w3.org/TR/wai-aria-1.1/#implicit_semantics)**) can be changed by attributes like aria-label and aria-labelledby?

**Comment by chrislachance** [comment]
I guess I'm asking that aria-label or aria-labelledby on role="row" functions exactly as role="rowheader" would if applied to cell 1 in that row.

As for what would happen if both exist at the same time, I'd imagine aria-label and aria-labelledby would win against role="rowheader" like they do in other contexts, but I'm not sure how the order of accessibility tree would be built out today with this conflict.

**Comment by chlane** [comment]
Thanks @chrislachance for further clarification are you asking for something like this to be added to the spec? -  "if an element has the implicit role of cell and is labelled by aria-label or aria-labelledby USER AGENTS MUST treat the element as if it is implemented with role='rowheader'."

**Comment by chrislachance** [comment]
Not quite, more like  "if an element has the implicit role of cell and its parent with the role of row is labelled by aria-label or aria-labelledby USER AGENTS MUST treat the row as if it is implemented with role='rowheader'."

**Comment by smhigley** [comment]
@chrislachance I think part of the problem is conflating what a `rowheader` is, and the name of the `role="row"` element. Both `aria-label` and `aria-labelledby` will name a `row`, but the row's name is not the same thing as the rowheader.

A `rowheader` is a child node of a row that has roughly the same structural requirements as a cell within a row. I can't really see a path to inserting a `rowheader` child node into a `row` based on adding an `aria-label` string, or coercing an element's role to become `rowheader` through `aria-labelledby` reference (which could easily point to an element outside the table/grid, or to an element that otherwise wouldn't support `rowheader` semantics). I'm happy to be corrected about that if anyone else has other thoughts on it.

Going through your examples, I think there are already existing methods to achieve what you describe, at least if I'm understanding your use cases:

> Simple issue example: First name column and Last name column aren't necessarily unique unto themselves, but the combination of the two can provide more context to data within that row.

For this, it is already possible to create multiple rowheaders, as well as multiple column headers: https://www.w3.org/WAI/tutorials/tables/multi-level/

> More complex issue example: First column is a row selection column, second is unique column data. Placing rowheader on column 2 isn't universally applied to the whole row in some screen readers, which kind of makes sense (NVDA assumes the header starts at column 2, JAWS does all items in row)

This seems like something that could be a logged as a bug on NVDA, rather than something that would change in the ARIA spec. If anything, Core AAM could potentially specify column/row header effects on cell accessibility API mappings, since that currently isn't covered, to my knowledge.

Another possibility -- I wasn't quite sure if what you're asking for is for screen readers to present the row's accessible name when a user navigates into it, independent of the existence of rowheaders. If that's the case, that would also be something to log with screen reader vendors. As the spec stands now, `row` can receive a name from author, so that seems like a reasonable ask.

I'm still not really sure I understood the ask, so feel free to correct the assumptions I used in this comment :)

**Comment by chrislachance** [comment]
I think what you are saying sounds right.

I guess what I'm expecting is that like I can use aria-labelledby on a Table element, instead of using a Caption, I'd be able to do the same for a row, replacing rowheader readback. There are often repeated actions in columns, and I'm hoping the context of those rows can be readback alongside those when entering the row (like entering a Table).

Looking at an example now in Codepen within Firefox & Chrome, I am seeing the row has a separate accessible label in the accessibility tree than the rowheader itself. However, not a single screen reader tested, JAWS, VoiceOver or NVDA are reading that row's heading back.

So what's next?

Looks like browsers are following the spec right, but its preventing context for the row to be read back to screen readers (tested VoiceOver, JAWS, and NVDA)

**Comment by sivakusayan** [comment]
I'm not sure how much the ARIA specification should care about compensating for user error, but I wonder how many websites in the wild place erroneous `aria-labels` on `row` elements. For example, [AG-Grid](https://www.ag-grid.com/example/), which seems to be a decently popular grid framework, gives each row `aria-label="Press Space to select row"`. Anecdotally, I also used to work for a company which gave extremely long `aria-labels` to each `row` in a `grid`. I'm not sure how common these issues are, but I wonder if this change would contribute to those issues by making table navigation excessively noisy.

>More complex issue example: First column is a row selection column, second is unique column data. Placing rowheader on column 2 isn't universally applied to the whole row in some screen readers, which kind of makes sense (NVDA assumes the header starts at column 2, JAWS does all items in row)

I do wish there was an accessible way to implement this pattern. I feel like one answer would be to say that a `rowheader` applies its semantics to all cells in that row, not just the cells to the right of that row. At least on the browser side:

- Chromium already has the above behavior (even if screen readers don't always respect it). However, I think it's technically breaking the [HTML spec](https://html.spec.whatwg.org/multipage/tables.html#the-th-element), and I'm not entirely sure this behavior is intentional.
- Firefox conforms to the HTML spec and only exposes `rowheader` semantics for cells to the right of the `rowheader`.

However, the answer I proposed could potentially break any table that tries to use multiple `rowheaders` in the same `row` to combine two or more tables into one table, such as [these examples](https://www.w3.org/WAI/tutorials/tables/multi-level/).
A safer way to get your example working is to finally have `headers` attribute support across browsers. Then, cells before the `rowheader` could have their `headers` attribute point to the correct rowheader. Unfortunately, the headers attribute is still poorly supported 😕

**Comment by scottaohara** [comment]
Yeah, ag-grid putting that aria-label on every row is a clear misunderstanding of what an aria-label is meant to represent.  As that text string is not a “name” but rather instructions for how to interact with an anti-pattern
